### PR TITLE
Adding Reader, Writer, and Rewriter to Trieste.

### PR DESCRIPTION
**Pull request by matajoh**
In the context of adding a means to convert YAML to JSON, this PR adds some new constructs to Trieste.
- A Reader is a generalized code construct which a language implementer can expose that parses a Source and returns an AST.
- A Writer is the same but for taking an AST and writing it to a Destination (also added in this PR)
- A Rewriter converts one AST to another AST via one or more rewrite passes.

In all three cases the implementer controls the activity of the construct via a series of rewrite passes. As part of this work, Driver was rewritten to optionally use a Reader, and the Process classes was refactored to make it a bit more flexible and easy to use.

Additionally, the YAML language implementation exposes a `to_json` converter, as well as a `event_writer` (for the existing event stream output) and a `writer` which emits YAML. All three are validated against the test suite. Both YAML and JSON expose Reader objects, and JSON exposes a Writer as well (a refactoring of its emitter).

### CODE REVIEW

File: include/trieste/passes.h [hunk 1]
```diff
@@ -112,65 +113,178 @@ namespace trieste
     std::chrono::microseconds duration;
   };
 
+  struct ProcessResult
+  {
+    bool ok;
+    std::string last_pass;
+    Node ast;
+    Node error;
```
Comment (mjp41) [inline include/trieste/passes.h:121]: Would this be better as a `Nodes`?  Then you don't need to introduce an `ErrorSeq` token?

File: include/trieste/passes.h [hunk 2]
```diff
@@ -216,120 +333,95 @@ namespace trieste
...
 
       wf::pop_front();
 
-      return ok;
+      return {ok, last_pass, ast, error};
     }
   };
 
-  inline void print_errors(Nodes& errors, logging::Log& err)
+  class Rewriter
```
Comment (mjp41) [inline include/trieste/passes.h:350]: Should this be in its own file?

**Review by mjp41 — APPROVED** [review]
LGTM.  Could of very minor things.

File: include/trieste/passes.h (hunk 2 above)
Comment (matajoh) [reply include/trieste/passes.h:350]: not a bad idea

**Review by mjp41 — APPROVED** [review]
LGTM. thanks

**Pull request merged**